lib/diamond/utils.ts

```diff
@@ -0,0 +1,305 @@
+        );
+
+        selectorsReplaced = true;
+      }
+    }
+  }
+
+  if (!selectorsReplaced) {
+    throw new Error('No selectors were replaced in FacetCut');
+  }
+
+  return groupFacetCuts(facetCuts);
+}
+
+// preview FacetCut which removes registered selectors
+export async function removeRegisteredSelectors(
+  diamond: IDiamondReadable,
+  contracts: Contract[],
+  only: FacetFilter[] = [],
+  except: FacetFilter[] = [],
+): Promise<FacetCut[]> {
+  validateFilters(only, except);
+
+  const diamondFacets: Facet[] = await diamond.facets();
+  const facets = getFacets(contracts);
+
+  let selectorsRemoved = false;
+  let facetCuts: FacetCut[] = [];
+
+  // if a registered selector is not found in the facets then it should be removed
+  // from the diamond
+  for (const diamondFacet of diamondFacets) {
+    for (const selector of diamondFacet.selectors) {
+      const target = diamondFacet.target;
+
+      if (
+        target != AddressZero &&
+        target != diamond.address &&
+        selector.length > 0 &&
+        !selectorExistsInFacets(selector, facets) &&
+        selectorIsFiltered(only, except, AddressZero, selector)
+      ) {
+        facetCuts.push(
+          printFacetCuts(AddressZero, [selector], FacetCutAction.REMOVE),
+        );
+
+        selectorsRemoved = true;
+      }
+    }
+  }
+
+  if (!selectorsRemoved) {
+    throw new Error('No selectors were removed from FacetCut');
+  }
+
+  return groupFacetCuts(facetCuts);
+}
+
+// preview a FacetCut which adds, replaces, or removes selectors, as needed
+export async function previewFacetCut(
+  diamond: IDiamondReadable,
+  contracts: Contract[],
+  only: FacetFilter[][] = [[], [], []],
+  except: FacetFilter[][] = [[], [], []],
+): Promise<FacetCut[]> {
+  let addFacetCuts: FacetCut[] = [];
+  let replaceFacetCuts: FacetCut[] = [];
+  let removeFacetCuts: FacetCut[] = [];
+
+  try {
+    addFacetCuts = await addUnregisteredSelectors(
+      diamond,
+      contracts,
+      only[0],
+      except[0],
+    );
+  } catch (error) {
+    console.log(`WARNING: ${(error as Error).message}`);
+  }
+
+  try {
+    replaceFacetCuts = await replaceRegisteredSelectors(
+      diamond,
+      contracts,
+      only[1],
+      except[1],
+    );
+  } catch (error) {
+    console.log(`WARNING: ${(error as Error).message}`);
+  }
+
+  try {
+    removeFacetCuts = await removeRegisteredSelectors(
+      diamond,
+      contracts,
+      only[2],
+      except[2],
+    );
+  } catch (error) {
+    console.log(`WARNING: ${(error as Error).message}`);
+  }
+
+  return groupFacetCuts([
+    ...addFacetCuts,
+    ...replaceFacetCuts,
+    ...removeFacetCuts,
+  ]);
+}
+
+// executes a DiamondCut using the provided FacetCut
+export async function diamondCut(
+  diamond: IDiamondWritable,
+  facetCut: FacetCut[],
+  target: string = AddressZero,
+  data: string = '0x',
+) {
+  (await diamond.diamondCut(facetCut, target, data)).wait(1);
```

Comment:
How exactly does `wait` work?

Reply:
> [Resolves to the TransactionReceipt once the transaction has been included in the chain for confirms blocks. If confirms is 0, and the transaction has not been mined, null is returned.](https://docs.ethers.org/v5/api/providers/types/#providers-TransactionResponse)

It may be good to return the TransactionReceipt here. Otherwise, a user may not be able to detect errors or access the receipt, if needed.

Reply:
https://github.com/solidstate-network/solidstate-solidity/pull/158/commits/871f8686765295d69347c3971a6517ad8c35d20d